# <center> Classification with PCA  </center>


## Loading  the libraries:

In [81]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt 
import seaborn as sns
sns.set(style="white")
sns.set(style="whitegrid", color_codes=True)
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report 
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier 
import lightgbm as lgb
from sklearn.ensemble import AdaBoostClassifier
import xgboost as xgb
from sklearn.ensemble import GradientBoostingClassifier
from catboost import CatBoostClassifier
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score


## Setting the Display option right:

In [82]:
def set_pandas_display_options() :

    pd.options.display.max_columns = 1000
    pd.options.display.max_rows = 1000
    pd.options.display.max_colwidth = 199
    pd.options.display.precision = 3

    
    
set_pandas_display_options()

## Loading the Dataset:

In [83]:
df = pd.read_csv('Downloads/features_phase6a.csv')

In [84]:
df.head()

FG1    FG2    FG3    FG4    FG5    FG6        FG7    FG8    FG9   FG10  \
0 -1.238  0.995 -0.737  1.077  0.254  0.082  1.719e-01  0.598 -0.437  0.030   
1 -1.441  0.582 -1.290  0.644 -0.383 -0.413 -1.231e+00 -0.345  0.769 -0.922   
2  0.000  0.000  0.000  0.000  0.000  0.000  0.000e+00  0.000  0.000  0.000   
3  2.024 -0.520 -1.129  0.263  0.185 -0.052 -1.958e-05 -0.152 -1.881  0.438   
4  0.905  0.130  1.092  0.525 -0.218 -0.223 -6.998e-01 -0.605  0.804 -0.315   

    FG11   FG12   FG13   FG14   FG15   FG16   FG17   FG18   FG19   FG20  \
0  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058 -0.232 -1.256  0.052   
1  1.059  0.358  0.330  0.119  0.076 -0.266  0.026  0.648 -0.191 -0.485   
2  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000   
3  0.133  0.893 -0.545  0.608  0.687 -1.822 -0.492  1.038 -0.477 -0.398   
4 -0.348  1.373  0.410 -0.487  0.029 -0.819 -1.698 -0.871  0.692  0.126   

    FG21   FG22   FG23   FG24   FG25   FG26   FG27   FG28   FG29   FG30  \
0 -0.696 -0.584  0.041  0.294  0.034 -1.728 -1.190 -0.835 -0.544 -0.167   
1  0.187 -0.154 -0.418  0.332 -0.876 -1.456 -0.437 -1.033 -1.728  0.067   
2  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000   
3 -0.322 -0.078 -0.401  0.100  0.840  0.393  1.127 -0.086 -1.790 -0.319   
4  0.446 -0.266  0.898  0.776  0.843 -0.505  0.873 -1.035 -0.196 -0.102   

    FG31   FG32   FG33   FG34   FG35   FG36   FG37   FG38   FG39   FG40  \
0  2.535  0.081 -0.418  0.309  0.219  0.048  0.549  0.898 -1.017 -1.596   
1  2.888  0.217 -0.362 -0.526 -0.452 -0.305  0.035 -0.500  0.670  0.186   
2  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000   
3  0.026  0.203 -1.040 -0.222 -0.607  0.227  0.249  1.236  0.544  0.477   
4  1.617 -0.327 -0.117 -0.603 -1.110  0.016 -0.272 -0.580 -0.411  0.028   

    FG41   FG42   FG43   FG44   FG45   FG46   FG47   FG48   FG49   FG50  \
0  0.483 -0.555 -0.244 -0.023 -0.911  0.150  0.145 -0.980  0.202 -0.361   
1 -0.456 -0.302  0.170  0.609  0.550  0.046 -0.999  0.614 -0.181  0.114   
2  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000   
3 -0.218  0.493 -0.737 -0.594 -0.448  0.410 -0.236 -0.372  1.498 -0.210   
4  2.288  0.070 -0.739  1.023  0.355 -0.214  0.452  0.628  0.347  0.463   

   SW_COM  SW_SEMCOL  SW_STOP  SW_COL  ND  LEN  DP_DOBJ  DP_PREP  DP_POBJ  \
0       0          0        0       0   0    2        0        0        0   
1       0          0        1       0   0   11        0        2        2   
2       7          0        1       0   8   18        0        0        0   
3       6          0        1       0   6   25        2        3        3   
4       1          0        2       1   0    2        0        0        0   

   DP_NSUBJ  DP_NLA  DP_NRA  DP_NNC    CG1    CG2    CG3    CG4    CG5    CG6  \
0         0   0.500   0.000   0.500 -1.238  0.995 -0.737  1.077  0.254  0.082   
1         0   0.333   0.583   0.333 -1.238  0.995 -0.737  1.077  0.254  0.082   
2         0   0.385   0.577   0.308 -1.238  0.995 -0.737  1.077  0.254  0.082   
3         1   0.375   0.562   0.250 -1.238  0.995 -0.737  1.077  0.254  0.082   
4         0   0.500   0.333   0.167 -1.238  0.995 -0.737  1.077  0.254  0.082   

     CG7    CG8    CG9  CG10   CG11   CG12   CG13   CG14   CG15   CG16   CG17  \
0  0.172  0.598 -0.437  0.03  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058   
1  0.172  0.598 -0.437  0.03  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058   
2  0.172  0.598 -0.437  0.03  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058   
3  0.172  0.598 -0.437  0.03  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058   
4  0.172  0.598 -0.437  0.03  0.288  0.423 -1.425 -0.416  1.113 -0.415 -0.058   

    CG18   CG19   CG20   CG21   CG22   CG23   CG24   CG25   CG26  CG27   CG28  \
0 -0.232 -1.256  0.052 -0.696 -0.584  0.041  0.294  0.034 -1.728 -1.19 -0.835   
1 -0.232 -1.256  0.052 -0.696 -0.584  0.041  0.294  0.034 -1.728 -1.19 -0.835   
2 -0.232 -1.256  0.052 -0.696 -0.584  0.041  0

In [86]:
df.shape

(17337, 135)

In [85]:
df.describe()

FG1        FG2        FG3        FG4        FG5        FG6  \
count  17337.000  17337.000  17337.000  17337.000  17337.000  17337.000   
mean       0.055      0.217     -0.019      0.123      0.108      0.132   
std        0.532      0.524      0.618      0.533      0.539      0.560   
min       -2.836     -2.194     -2.466     -2.702     -2.462     -2.889   
25%       -0.237      0.000     -0.465     -0.242     -0.218     -0.152   
50%        0.000      0.130      0.000      0.000      0.082      0.023   
75%        0.418      0.571      0.349      0.525      0.499      0.527   
max        2.908      2.472      2.223      3.135      2.318      3.078   

             FG7        FG8        FG9       FG10       FG11       FG12  \
count  17337.000  17337.000  17337.000  17337.000  17337.000  17337.000   
mean      -0.048     -0.349     -0.066     -0.024      0.059      0.081   
std        0.617      0.580      0.647      0.542      0.449      0.566   
min       -2.311     -2.553     -2.380     -2.272     -1.733     -2.861   
25%       -0.497     -0.674     -0.354     -0.315     -0.252     -0.156   
50%       -0.009     -0.263      0.000      0.000      0.000      0.000   
75%        0.281      0.000      0.230      0.175      0.310      0.438   
max        2.495      2.378      3.311      2.765      2.108      2.086   

            FG13       FG14       FG15       FG16       FG17       FG18  \
count  17337.000  17337.000  17337.000  17337.000  17337.000  17337.000   
mean      -0.108     -0.118      0.113     -0.096     -0.230     -0.170   
std        0.504      0.453      0.507      0.456      0.625      0.501   
min       -2.228     -2.243     -2.480     -2.482     -2.643     -1.991   
25%       -0.414     -0.416     -0.077     -0.415     -0.519     -0.453   
50%        0.000     -0.073      0.029     -0.015     -0.058     -0.158   
75%        0.179      0.116      0.367      0.131      0.155      0.023   
max        2.295      2.343      1.845      1.812      1.977      2.284   

            FG19       FG20       FG21       FG22       FG23       FG24  \
count  17337.000  17337.000  17337.000  17337.000  17337.000  17337.000   
mean      -0.137     -0.085      0.124     -0.116      0.195      0.108   
std        0.678      0.605      0.493      0.548      0.603      0.477   
min       -2.300     -2.575     -2.679     -2.467     -2.014     -2.071   
25%       -0.495     -0.412     -0.022     -0.448     -0.204     -0.185   
50%        0.000     -0.012      0.052     -0.022      0.009      0.000   
75%        0.370      0.217      0.390      0.161      0.712      0.376   
max        2.049      2.536      2.554      1.927      2.635      2.764   

            FG25       FG26       FG27       FG28       FG29       FG30  \
count  17337.000  17337.000  17337.000  17337.000  17337.000  17337.000   
mean       0.038     -0.598     -0.161     -0.173     -0.080     -0.035   
std        0.549      0.816      0.658      0.450      0.611      0.501   
min       -2.383     -2.796     -2.384     -1.823     -2.914     -2.152   
25%       -0.250     -1.090     -0.586     -0.447     -0.419     -0.282   
50%        0.000     -0.548     -0.090     -0.072      0.000      0.000   
75%        0.348      0.000      0.244      0.099      0.348      0.265   
max        2.430      2.141      2.310      2.047      1.839      2.058   

            FG31       FG32       FG33       FG34       FG35       FG36  \
count  17337.000  17337.000  17337.000  17337.000  1.734e+04  17337.000   
mean       1.725     -0.160     -0.114     -0.355  5.419e-02     -0.068   
std        1.459      0.561      0.446      0.768  4.901e-01      0.424   
min       -2.243     -2.455     -1.896     -2.749 -2.533e+00     -2.005   
25%        0.201     -0.466     -0.396     -0.822 -1.607e-01     -0.269   
50%        1.716     -0.073     -0.117     -0.231  5.921e-04      0.000   
75%        2.991      0.207      0.050      0.025  3.081e-01      0.106   
max        4.247      1.786      

In [58]:
df.isnull().sum()

FG1          0
FG2          0
FG3          0
FG4          0
FG5          0
FG6          0
FG7          0
FG8          0
FG9          0
FG10         0
FG11         0
FG12         0
FG13         0
FG14         0
FG15         0
FG16         0
FG17         0
FG18         0
FG19         0
FG20         0
FG21         0
FG22         0
FG23         0
FG24         0
FG25         0
FG26         0
FG27         0
FG28         0
FG29         0
FG30         0
FG31         0
FG32         0
FG33         0
FG34         0
FG35         0
FG36         0
FG37         0
FG38         0
FG39         0
FG40         0
FG41         0
FG42         0
FG43         0
FG44         0
FG45         0
FG46         0
FG47         0
FG48         0
FG49         0
FG50         0
SW_COM       0
SW_SEMCOL    0
SW_STOP      0
SW_COL       0
ND           0
LEN          0
DP_DOBJ      0
DP_PREP      0
DP_POBJ      0
DP_NSUBJ     0
DP_NLA       0
DP_NRA       0
DP_NNC       0
CG1          0
CG2          0
CG3          0
CG4       

<div class="alert alert-block alert-info"><b> 
    
-  The dataset have 135 columns & 17337 rows.
-  Except the target variables are the variables are numerical.
-  There is no zero importance feature present in the data
-  There is no missing values in the data set
</b></div>    


In [59]:
df['LABEL'].value_counts()

affiliation              3704
correspondingauthor      3138
abstract                 2604
Others                   1821
authors                  1678
keywordsdefault          1487
articletitle             1283
abstracttitle             615
dummy                     418
acknowledgements          145
listdefinition            128
articlefootnote           100
grant-sponsor              92
grant-highlight            91
acknowledgementstitle      22
nomenclature                9
grant-number                2
Name: LABEL, dtype: int64

## Seperating the variables:

In [89]:
X = df.iloc[:, 0:134].values 
y = df.iloc[:, 134].values 

## Train Test Split:

In [90]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2 , random_state = 123)

## Standardization:

In [91]:
from sklearn.preprocessing import StandardScaler 
sc = StandardScaler() 
  
X_train = sc.fit_transform(X_train) 
X_test = sc.transform(X_test) 

## PCA:

In [92]:
from sklearn.decomposition import PCA 
  
pca = PCA(n_components = 30) 
  
X_train = pca.fit_transform(X_train) 
X_test = pca.transform(X_test) 
  
explained_variance = pca.explained_variance_ratio_

<div class="alert alert-block alert-info"><b> 
    
-  I have used 30 principal components .
-  Means from 134 variables the varianace will be explained by 30 variables.
</b></div>

## Model Evaluation:

In [ ]:
def model_evaluation(actuals, predictions) : 
    
    print('\nReport : ')
    print(classification_report(actuals, predictions))
    
    print('\nAccuracy Score :', accuracy_score(actuals, predictions)) 
    print("------------------------------------------------------------------")

## Model Building:

## Logistic Regression:

In [94]:
logreg = LogisticRegression()
logreg.fit(X_train, y_train)
pred_train_logreg = logreg.predict(X_train)
pred_test_logreg = logreg.predict(X_test)

In [95]:
model_evaluation(y_train , pred_train_logreg)



Report : 
                       precision    recall  f1-score   support

               Others       0.69      0.67      0.68      1460
             abstract       0.86      0.85      0.85      2083
        abstracttitle       0.80      0.97      0.87       504
     acknowledgements       0.39      0.06      0.11       108
acknowledgementstitle       0.90      0.50      0.64        18
          affiliation       0.78      0.86      0.82      2928
      articlefootnote       0.00      0.00      0.00        89
         articletitle       0.72      0.84      0.78      1051
              authors       0.69      0.71      0.70      1361
  correspondingauthor       0.67      0.72      0.70      2486
                dummy       0.00      0.00      0.00       316
      grant-highlight       0.43      0.11      0.18        79
         grant-number       0.00      0.00      0.00         2
        grant-sponsor       0.00      0.00      0.00        76
      keywordsdefault       0.80      0.78 

In [96]:
model_evaluation(y_test , pred_test_logreg)



Report : 
                       precision    recall  f1-score   support

               Others       0.70      0.66      0.68       361
             abstract       0.86      0.85      0.85       521
        abstracttitle       0.80      0.95      0.87       111
     acknowledgements       0.33      0.05      0.09        37
acknowledgementstitle       1.00      0.25      0.40         4
          affiliation       0.79      0.88      0.83       776
      articlefootnote       0.00      0.00      0.00        11
         articletitle       0.64      0.81      0.72       232
              authors       0.66      0.72      0.69       317
  correspondingauthor       0.69      0.73      0.71       652
                dummy       0.00      0.00      0.00       102
      grant-highlight       0.00      0.00      0.00        12
        grant-sponsor       0.33      0.06      0.11        16
      keywordsdefault       0.79      0.74      0.76       287
       listdefinition       0.00      0.00 

## Decision Tree:

In [97]:
dtc = DecisionTreeClassifier()
dtc.fit(X_train, y_train)
pred_train_dtc = dtc.predict(X_train)
pred_test_dtc = dtc.predict(X_test)

In [98]:
model_evaluation(y_train , pred_train_dtc)



Report : 
                       precision    recall  f1-score   support

               Others       0.94      0.95      0.94      1460
             abstract       1.00      0.98      0.99      2083
        abstracttitle       0.86      0.98      0.92       504
     acknowledgements       0.99      0.93      0.96       108
acknowledgementstitle       1.00      1.00      1.00        18
          affiliation       1.00      1.00      1.00      2928
      articlefootnote       1.00      0.85      0.92        89
         articletitle       1.00      1.00      1.00      1051
              authors       1.00      1.00      1.00      1361
  correspondingauthor       0.99      0.99      0.99      2486
                dummy       0.98      0.89      0.93       316
      grant-highlight       1.00      0.96      0.98        79
         grant-number       1.00      1.00      1.00         2
        grant-sponsor       1.00      0.96      0.98        76
      keywordsdefault       1.00      1.00 

In [99]:
model_evaluation(y_test , pred_test_dtc)



Report : 
                       precision    recall  f1-score   support

               Others       0.73      0.71      0.72       361
             abstract       0.84      0.83      0.84       521
        abstracttitle       0.81      0.95      0.88       111
     acknowledgements       0.43      0.35      0.39        37
acknowledgementstitle       0.00      0.00      0.00         4
          affiliation       0.83      0.81      0.82       776
      articlefootnote       0.00      0.00      0.00        11
         articletitle       0.69      0.70      0.70       232
              authors       0.69      0.71      0.70       317
  correspondingauthor       0.75      0.76      0.75       652
                dummy       0.37      0.30      0.34       102
      grant-highlight       0.17      0.33      0.22        12
        grant-sponsor       0.45      0.31      0.37        16
      keywordsdefault       0.78      0.74      0.75       287
       listdefinition       0.63      0.59 

## Random Forest:

In [100]:
rf = RandomForestClassifier()
rf.fit(X_train, y_train)
pred_train_rf = rf.predict(X_train)
pred_test_rf = rf.predict(X_test)

In [101]:
model_evaluation(y_train , pred_train_rf)


Report : 
                       precision    recall  f1-score   support

               Others       0.95      0.93      0.94      1460
             abstract       0.99      0.98      0.98      2083
        abstracttitle       0.86      0.99      0.92       504
     acknowledgements       0.96      0.91      0.93       108
acknowledgementstitle       1.00      1.00      1.00        18
          affiliation       0.99      1.00      0.99      2928
      articlefootnote       0.96      0.84      0.90        89
         articletitle       0.99      0.99      0.99      1051
              authors       0.99      0.99      0.99      1361
  correspondingauthor       0.99      0.98      0.98      2486
                dummy       0.95      0.87      0.91       316
      grant-highlight       1.00      0.89      0.94        79
         grant-number       1.00      1.00      1.00         2
        grant-sponsor       0.99      0.96      0.97        76
      keywordsdefault       0.99      0.99 

In [102]:
model_evaluation(y_test , pred_test_rf)



Report : 
                       precision    recall  f1-score   support

               Others       0.74      0.77      0.76       361
             abstract       0.90      0.89      0.89       521
        abstracttitle       0.83      0.96      0.89       111
     acknowledgements       0.75      0.32      0.45        37
acknowledgementstitle       1.00      0.25      0.40         4
          affiliation       0.81      0.90      0.85       776
      articlefootnote       0.00      0.00      0.00        11
         articletitle       0.71      0.83      0.77       232
              authors       0.76      0.77      0.76       317
  correspondingauthor       0.82      0.78      0.80       652
                dummy       0.63      0.26      0.37       102
      grant-highlight       0.50      0.25      0.33        12
        grant-sponsor       0.86      0.38      0.52        16
      keywordsdefault       0.82      0.78      0.80       287
       listdefinition       0.72      0.45 

## XG Boost:

In [103]:
xg = xgb.XGBClassifier()
xg.fit(X_train, y_train)
pred_train_xg = xg.predict(X_train)
pred_test_xg = xg.predict(X_test)

In [104]:
model_evaluation(y_train , pred_train_xg)


Report : 
                       precision    recall  f1-score   support

               Others       0.78      0.75      0.77      1460
             abstract       0.92      0.88      0.90      2083
        abstracttitle       0.84      0.99      0.91       504
     acknowledgements       0.78      0.35      0.48       108
acknowledgementstitle       0.94      0.94      0.94        18
          affiliation       0.81      0.89      0.85      2928
      articlefootnote       1.00      0.27      0.42        89
         articletitle       0.77      0.88      0.82      1051
              authors       0.77      0.78      0.78      1361
  correspondingauthor       0.77      0.80      0.78      2486
                dummy       0.80      0.22      0.35       316
      grant-highlight       0.96      0.33      0.49        79
         grant-number       1.00      1.00      1.00         2
        grant-sponsor       1.00      0.38      0.55        76
      keywordsdefault       0.86      0.86 

In [105]:
model_evaluation(y_test , pred_test_xg)


Report : 
                       precision    recall  f1-score   support

               Others       0.74      0.72      0.73       361
             abstract       0.89      0.87      0.88       521
        abstracttitle       0.82      0.98      0.89       111
     acknowledgements       0.50      0.11      0.18        37
acknowledgementstitle       0.67      0.50      0.57         4
          affiliation       0.79      0.88      0.83       776
      articlefootnote       0.00      0.00      0.00        11
         articletitle       0.66      0.82      0.73       232
              authors       0.73      0.73      0.73       317
  correspondingauthor       0.75      0.78      0.76       652
                dummy       0.58      0.07      0.12       102
      grant-highlight       0.50      0.25      0.33        12
        grant-sponsor       0.80      0.25      0.38        16
      keywordsdefault       0.82      0.76      0.79       287
       listdefinition       0.75      0.41 

# Conclusion:

<div class="alert alert-block alert-info"><b> 
    
-  After applying PCA with 30 componenets the accuracy is still the same.
-  The best classiifer is XGBoost.
-  Tree Classifier are overfitting.
-  I have also done Grid Search CV to hypertune the tree classifiers.
</b></div>